File: vanish.py
import sys
import requests
import argparse
import subprocess
from concurrent.futures import ThreadPoolExecutor, as_completed
from fake_useragent import UserAgent

# List of CDN providers to check in HTTP headers and DNS results
CDN_PROVIDERS = ["akamai", "imperva", "cloudflare", "fastly", "verizon", "stackpath", "incapsula"]

# Random user-agent generator
ua = UserAgent()

def random_headers():
    """Generate random HTTP headers including a random User-Agent."""
    return {"User-Agent": ua.random}

def check_http(domain, verbose=False):
    """Check if the domain is behind a CDN by analyzing HTTP headers."""
    try:
        response = requests.get(f"http://{domain}", headers=random_headers(), timeout=5)
        headers = response.headers
        for provider in CDN_PROVIDERS:
            if provider in headers.get("server", "").lower() or provider in headers.get("x-cdn", "").lower():
                if verbose:
                    print(f"{domain} (CDN detected via HTTP: {provider.capitalize()})")
                return None
        return domain
    except requests.exceptions.RequestException:
        return None

def check_dns(domain, verbose=False):
    """Check DNS resolution using both 'host' and 'dig' commands for CDN detection."""
    try:
        # Run 'host' command to get DNS resolution
        result_host = subprocess.run(['host', domain], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        output_host = result_host.stdout.decode()

        # Check 'host' output for CDN providers
        for provider in CDN_PROVIDERS:
            if provider in output_host.lower():
                if verbose:
                    print(f"{domain} (CDN detected via DNS 'host': {provider.capitalize()})")
                return None

        # Run 'dig' command to check DNS resolution for CDN providers
        result_dig = subprocess.run(['dig', domain], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        output_dig = result_dig.stdout.decode()

        for provider in CDN_PROVIDERS:
            if provider in output_dig.lower():
                if verbose:
                    print(f"{domain} (CDN detected via DNS 'dig': {provider.capitalize()})")
                return None

        return domain
    except subprocess.SubprocessError:
        return None

def process_domain(domain, verbose=False):
    """Process a single domain by checking both DNS and HTTP for CDN presence."""
    # Remove the ":port" part if it exists
    if ':' in domain:
        domain = domain.split(':')[0]

    # First check via DNS resolution using 'host' and 'dig'
    domain_check = check_dns(domain, verbose)
    if domain_check:
        # If DNS is clear, check via HTTP headers
        return check_http(domain, verbose)
    return None

def main():
    """Main function to parse arguments and execute CDN detection."""
    parser = argparse.ArgumentParser(description="Vanish: Filter domains behind CDN (Akamai, Imperva, Cloudflare, etc.)")
    parser.add_argument("-l", "--list", help="File containing the list of domains")
    parser.add_argument("-t", "--target", help="Check a single domain")
    parser.add_argument("-c", "--concurrence", type=int, default=10, help="Number of threads for parallel execution")
    parser.add_argument("-v", "--verbose", action="store_true", help="Enable verbose mode to show CDN detections")
    args = parser.parse_args()

    domains = []

    if args.list:
        # Read domains from a file
        with open(args.list, 'r') as file:
            domains = [line.strip() for line in file.readlines()]
    elif args.target:
        # Single domain via argument
        domains = [args.target]
    else:
        # Read domains from stdin
        domains = [line.strip() for line in sys.stdin.readlines()]

    # Run checks in parallel using the specified number of threads
    with ThreadPoolExecutor(max_workers=args.concurrence) as executor:
        futures = {executor.submit(process_domain, domain, args.verbose): domain for domain in domains}

        for future in as_completed(futures):
            result = future.result()
            if result:
                print(result)

if __name__ == "__main__":
    main()

File: vanish.go
package main

import (
    "bufio"
    "flag"
    "fmt"
    "io"
    "log"
    "math/rand"
    "net/http"
    "os"
    "os/exec"
    "strings"
    "sync"
    "time"
)

// Lista de provedores de CDN/WAF
var cdnProviders = []string{
    "akamai",
    "imperva",
    "cloudflare",
    "fastly",
    "verizon",
    "stackpath",
    "incapsula",
}

// Gera um User-Agent aleatório simples
func randomUserAgent() string {
    agents := []string{
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64)",
        "Mozilla/5.0 (X11; Linux x86_64)",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7)",
        "Mozilla/5.0 (iPhone; CPU iPhone OS 14_2 like Mac OS X)",
        "Mozilla/5.0 (Linux; Android 10)",
    }
    rand.Seed(time.Now().UnixNano())
    return agents[rand.Intn(len(agents))]
}

// checkDNS usa os comandos "host" e "dig" para detectar se um domínio está atrás de um WAF/CDN.
// Retorna true se DETECTOU CDN (ou seja, deve ser filtrado), senão false.
func checkDNS(domain string, verbose bool) bool {
    // Comando "host"
    outputHost, err := exec.Command("host", domain).CombinedOutput()
    if err == nil {
        hostString := strings.ToLower(string(outputHost))
        for _, provider := range cdnProviders {
            if strings.Contains(hostString, provider) {
                if verbose {
                    fmt.Printf("%s (CDN detectado via DNS 'host': %s)\n", domain, strings.Title(provider))
                }
                return true // CDN detectado
            }
        }
    }

    // Comando "dig"
    outputDig, err := exec.Command("dig", domain).CombinedOutput()
    if err == nil {
        digString := strings.ToLower(string(outputDig))
        for _, provider := range cdnProviders {
            if strings.Contains(digString, provider) {
                if verbose {
                    fmt.Printf("%s (CDN detectado via DNS 'dig': %s)\n", domain, strings.Title(provider))
                }
                return true
            }
        }
    }

    return false
}

// checkHTTP faz uma requisição HTTP e verifica os cabeçalhos "Server" e "X-CDN".
// Retorna true se DETECTOU CDN, senão false.
func checkHTTP(domain string, verbose bool) bool {
    client := http.Client{
        Timeout: 5 * time.Second,
    }

    url := "http://" + domain
    req, err := http.NewRequest("GET", url, nil)
    if err != nil {
        return false // não conseguiu criar a requisição, assume não detectado
    }
    req.Header.Set("User-Agent", randomUserAgent())

    resp, err := client.Do(req)
    if err != nil {
        // Erro de conexão ou timeout; não necessariamente indica WAF, então retornamos false
        return false
    }
    defer resp.Body.Close()

    serverHeader := strings.ToLower(resp.Header.Get("Server"))
    xcdnHeader := strings.ToLower(resp.Header.Get("X-CDN"))

    for _, provider := range cdnProviders {
        if strings.Contains(serverHeader, provider) || strings.Contains(xcdnHeader, provider) {
            if verbose {
                fmt.Printf("%s (CDN detectado via HTTP: %s)\n", domain, strings.Title(provider))
            }
            return true
        }
    }

    return false
}

// processDomain faz o fluxo principal de checagem: DNS -> HTTP.
// Retorna true se DOMÍNIO ESTÁ LIMPO (não está atrás de WAF/CDN), ou false se deve ser filtrado.
func processDomain(domain string, verbose bool) bool {
    // Remover ":port" se existir
    if strings.Contains(domain, ":") {
        domain = strings.Split(domain, ":")[0]
    }

    // Primeiro checa via DNS
    if checkDNS(domain, verbose) {
        return false
    }

    // Se não detectou via DNS, checa via HTTP
    if checkHTTP(domain, verbose) {
        return false
    }

    // Se chegou até aqui, não detectou nenhum WAF/CDN
    return true
}

// readDomains auxilia a ler os domínios de arquivo, target ou stdin
func readDomains(listFile, target string) ([]string, error) {
    var domains []string

    switch {
    case listFile != "":
        f, err := os.Open(listFile)
        if err != nil {
            return nil, fmt.Errorf("erro ao abrir arquivo: %v", err)
        }
        defer f.Close()

        scanner := bufio.NewScanner(f)
        for scanner.Scan() {
            line := strings.TrimSpace(scanner.Text())
            if line != "" {
                domains = append(domains, line)
            }
        }
        if err := scanner.Err(); err != nil {
            return nil, err
        }

    case target != "":
        domains = []string{target}

    default:
        // Lê de stdin
        info, err := os.Stdin.Stat()
        if err != nil {
            return nil, err
        }
        if (info.Mode() & os.ModeCharDevice) != 0 {
            return nil, fmt.Errorf("nenhum input de arquivo/lista, target ou stdin")
        }

        reader := bufio.NewReader(os.Stdin)
        for {
            line, err := reader.ReadString('\n')
            if err != nil && err != io.EOF {
                return nil, err
            }
            line = strings.TrimSpace(line)
            if line != "" {
                domains = append(domains, line)
            }
            if err == io.EOF {
                break
            }
        }
    }

    return domains, nil
}

func main() {
    var (
        listFile   string
        singleTgt  string
        concurrency int
        verbose    bool
    )

    flag.StringVar(&listFile, "l", "", "Arquivo contendo a lista de domínios")
    flag.StringVar(&singleTgt, "t", "", "Domínio único para checar")
    flag.IntVar(&concurrency, "c", 10, "Número de threads (goroutines) para execução em paralelo")
    flag.BoolVar(&verbose, "v", false, "Exibir modo verboso (mostrar detecções de WAF/CDN)")
    flag.Parse()

    // Lê domínios de acordo com os parâmetros
    domains, err := readDomains(listFile, singleTgt)
    if err != nil {
        log.Fatalf("Erro ao ler domínios: %v\n", err)
    }

    // Canal para enfileirar os domínios
    domainChan := make(chan string)
    var wg sync.WaitGroup

    // Cria workers
    for i := 0; i < concurrency; i++ {
        wg.Add(1)
        go func() {
            defer wg.Done()
            for d := range domainChan {
                if processDomain(d, verbose) {
                    // Se não está por trás de CDN, imprime o domínio
                    fmt.Println(d)
                }
            }
        }()
    }

    // Envia domínios para o canal
    go func() {
        for _, d := range domains {
            domainChan <- d
        }
        close(domainChan)
    }()

    // Aguarda todos terminarem
    wg.Wait()
}
